### PR TITLE
Docs/streams

### PR DESCRIPTION
**Pull request by vtjeng**
**Goals (and why)**: Update description of streams in documentation.

**Implementation Description (bullets)**: N/A

**Concerns (what feedback would you like?)**: Accuracy of description of streams API, and ensuring that I didn't remove any important descriptions.

**Where should we start reviewing?**: A read through of the single changed file should do.

**Priority (whenever / two weeks / yesterday)**: Two weeks.

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/2227)

### CODE REVIEW

**Comment by palantirtech** [comment]
Thanks for your interest in [palantir/atlasdb](https://github.com/palantir/atlasdb), @vtjeng! Before we can accept your pull request, you need to sign our contributor license agreement - just visit https://cla.palantir.com/ and follow the instructions. Once you sign, I'll automatically update this pull request.

File: docs/source/schemas/streams.rst [hunk 1]
```diff
@@ -10,32 +10,61 @@ Overview
...
+        new StreamStoreDefinitionBuilder(String shortName, String longName, ValueType valueType).build()
+    )
+    
+- ``shortName`` is the prefix which will be used in the actual database tables.
+- ``longName`` is the name of the stream store in the generated Java code.
+- ``valueType`` specifies the value type of the id for each stored object.
+
+Additional options for the builder include:
+
+- ``hashFirstRowComponent``
```
Comment (gsheasby) [inline docs/source/schemas/streams.rst]: This would probably look nicer as a simple table.

File: docs/source/schemas/streams.rst [hunk 2]
```diff
@@ -10,32 +10,61 @@ Overview
...
+  stream via the LZ4 algorithm upon reads and writes, respectively. 
+  Compression is performed client side before any network communication to 
+  the underlying database.
+- ``inMemoryThreshold(int inMemoryThreshold)``: specifies the largest size
+  object (in bytes) which AtlasDB will cache in memory in order to boost
+  retrieval performance.
+  
+For an example of streams in use, see `ProfileSchema`_ and `ProfileStore`_.
+
+.. _ProfileSchema: ../../../examples/profile-client/src/main/java/com/palantir/example/profile/schema/ProfileSchema.java
```
Comment (gsheasby) [inline docs/source/schemas/streams.rst]: This is risky - ProfileSchema/Store might change. We could instead copy code snippets (if they're small), or link to a particular blob in public github (e.g. the 0.52.0 version of these files).

File: docs/source/schemas/streams.rst [hunk 3]
```diff
@@ -10,32 +10,61 @@ Overview
...
+2. ``storeStream(InputStream stream)`` is called *outside* a transaction. It will store a stream without marking it as used, and users should mark the stream as used via ``markStreamAsUsed(Transaction t, long streamId, byte[] reference)`` immediately after. Users of this method need to retry if their stream was cleaned up between storing and marking. This method is the preferred method for very large streams (>100 MB) that would otherwise cause the transaction within which the stream is being stored to be too long running.
+
+Regular AtlasDB tables should refer to the object stored in stream stores by their ``streamId``, and the ``reference`` that users mark the stream with should correspond to a value computable based on the parameters of the cell the ``streamId`` is stored in. If the ``streamId`` is deleted from a cell, users need to mark the stream as unused by the reference corresponding to that cell to enable garbage collection.
+
+.. note::
+    
+    References to streams are *manually* added and removed by users. AtlasDB monitors the number of references to each stream, and streams that are unreferenced will be garbage collected.
```
Comment (gsheasby) [inline docs/source/schemas/streams.rst]: This would be a good place to mention `markStreamAsUnused`.

File: docs/source/schemas/streams.rst (hunk 1 above)
Comment (gsheasby) [inline docs/source/schemas/streams.rst]: Also, we should note what these first three mean:
* `hashFirstRowComponent`: hashes the "id" component of the table
* `isAppendHeavyAndReadLight`: `false` by default. If set to `true`, some optimisations will be made in Cassandra KVS assuming that these streams will be changed often but not read often.
* `compressBlocksInDb` ... this is never actually used, AFAICT. Let's remove this reference from the docs.

**Comment by vtjeng** [comment]
Thanks for the review. I'd like to own continuing making changes. 

For the code example, I decided to link to a particular example use rather than copying the code snippets since these snippets are relatively large.

File: docs/source/schemas/streams.rst [hunk 4]
```diff
@@ -32,22 +32,29 @@ The minimal parameters required to define a stream store are shown below:
...
+    *    - ``compressStreamInClient``
+         - Transparently decompresses and compresses the stream via the LZ4 algorithm upon reads and writes, respectively. Compression is performed client side before any network communication to the underlying database.
+
+    *   - ``inMemoryThreshold``
+        - Specifies the largest size object (in bytes) which AtlasDB will cache in memory in order to boost retrieval performance.
+  
+For an example of streams in use, see the ``user_profile`` table and ``user_photos`` stream store in `ProfileSchema`_, and the ``updateImage`` method in `ProfileStore`_.
 
+.. _ProfileSchema: https://github.com/palantir/atlasdb/blob/cd4f33dfcaa95acb90374f698158a4aae8c28945/examples/profile-client/src/main/java/com/palantir/example/profile/schema/ProfileSchema.java
+.. _ProfileStore: https://github.com/palantir/atlasdb/blob/cd4f33dfcaa95acb90374f698158a4aae8c28945/examples/profile-client/src/main/java/com/palantir/example/profile/ProfileStore.java
```
Comment (gsheasby) [inline docs/source/schemas/streams.rst]: You can add `#L92` to this link, to link the viewer directly to the relevant part of the file.

**Pull request merged**